packages/venia-concept/src/components/SearchBar/searchBar.js

```diff
@@ -1,173 +1,82 @@
-        const { resetForm, state } = this;
-
-        return state.dirty && <Trigger action={resetForm}>{clearIcon}</Trigger>;
-    }
-
-    updateAutocompleteVisible = visible => {
-        this.setState({
-            autocompleteVisible: visible
-        });
-    };
-
-    handleChange = ({ values }) => {
-        const dirty = !!values.search_query;
-        dirty !== this.state.dirty && this.setState({ dirty });
-    };
-
-    handleInputChange = () => this.updateAutocompleteVisible(true);
-
-    handleSubmit = ({ search_query }) => {
-        if (search_query) {
-            this.props.executeSearch(search_query, this.props.history);
-            this.updateAutocompleteVisible(false);
-        }
-    };
-
-    resetForm = () => {
-        this.updateAutocompleteVisible(false);
-        this.formApi.reset();
-    };
-
-    setApi = formApi => {
-        this.formApi = formApi;
-    };
-
-    render() {
-        const { props, resetButton, formApi } = this;
-        const { autocompleteVisible } = this.state;
-        const { classes, isOpen } = props;
-        const className = isOpen ? classes.root_open : classes.root;
-
-        const { values } = (formApi && formApi.getState()) || {};
-        const searchQuery = (values && values.search_query) || '';
-
-        return (
-            <div className={className}>
-                <div className={classes.searchInner} ref={this.searchRef}>
-                    <Form
-                        className={classes.form}
-                        getApi={this.setApi}
-                        autoComplete="off"
-                        initialValues={initialValues}
-                        onChange={this.handleChange}
-                        onSubmit={this.handleSubmit}
-                    >
-                        <TextInput
-                            field="search_query"
-                            onFocus={this.inputFocus}
-                            onChange={this.handleInputChange}
-                            after={resetButton}
-                            before={searchIcon}
+const initialValues = { search_query: '' };
+
+const SearchBar = props => {
+    const { history, isOpen, location } = props;
+    const { elementRef, expanded, setExpanded } = useDropdown();
+
+    const classes = mergeClasses(defaultClasses, props.classes);
+    const rootClassName = isOpen ? classes.root_open : classes.root;
+
+    // expand or collapse on input change
+    const handleChange = useCallback(value => {
+        setExpanded(!!value);
+    }, []);
+
+    // expand on focus
+    const handleFocus = useCallback(() => {
+        setExpanded(true);
+    }, []);
+
+    // navigate on submit
+    const handleSubmit = useCallback(
+        ({ search_query }) => {
+            history.push(`/search.html?query=${search_query}`);
+        },
+        [history]
+    );
+
+    return (
+        <div className={rootClassName}>
+            <div ref={elementRef} className={classes.container}>
```

Comment:
There's still a lot of debate in the React community about whether custom hooks should return arrays or objects, but suffice to say the justification for returning an array is actually very thin. One of my goals with custom hooks is to encourage exposing as much as possible (over-returning) and to make it easy for consumers to just use the relevant subset of the return value. So I'll prefer objects for our custom hooks.

In general, the only purpose of a custom hook is to save you the trouble of orchestrating built-in hook calls yourself, so here I figured `useDropdown` might as well call `useRef` instead of making the consumer do it. However, I could imagine a scenario where a consumer wanted to customize the ref (with methods, etc.) and would prefer to create it and pass it in. I'll try to revise this.

Reply:
> There's still a lot of debate in the React community about whether custom hooks should return arrays or objects, but suffice to say the justification for returning an array is actually very thin.

Yeah, not forcing your naming on the consumer is really the only thing I can think of.

> One of my goals with custom hooks is to encourage exposing as much as possible (over-returning) and to make it easy for consumers to just use the relevant subset of the return value. So I'll prefer objects for our custom hooks.

Totally dig it, great point 👍

Reply:
> Yeah, not forcing your naming on the consumer is really the only thing I can think of.

Right. Arrays and objects each offer sweeter syntax in different cases.

```es6
const useDropdown = () => {
  const elementRef = useRef(null)
  const [expanded, setExpanded] = useState(false)

  // return array or object?
  return
}

// array return
// easy to rename things, but harder to destructure a subset
const [_, visible, setVisible] = useDropdown()

// object return
// easy to destructure a subset, but harder to rename things
const { expanded: visible, setExpanded: setVisible } = useDropdown()
```

Reply:
I say return objects. Handling parameter order in untyped JavaScript is hard to deal with, compared to aliased object destructuring. It's easy to rename an object property. It's harder to deal with an opaque TypeError or ReferenceError when you got your order wrong.